docs/root/configuration/listener_filters/original_src_filter.rst

```diff
@@ -0,0 +1,75 @@
+.. _config_listener_filters_original_src:
+
+Original Source
+===============
+
+* :ref:`Listener filter v2 API reference <envoy_api_msg_config.filter.listener.original_src.v2alpha1.OriginalSrc>`
+* This filter should be configured with the name *envoy.listener.original_src*.
+
+The original source listener filter replicates the downstream remote address of the connection on
+the upstream side of Envoy. For example, if a downstream connection connects to Envoy with IP
+address ``10.1.2.3``, then Envoy will connect to the upstream with source IP ``10.1.2.3``.
+
+Interaction with Proxy Protocol
+--------------------------------
+
+If the connection has not had its source address translated or proxied, then Envoy can simply use
+the existing connection information to build the correct downstream remote address. However, if this
+is not true, a :ref:`Proxy Protocol filter <config_listener_filters_proxy_protocol>` may be used to
+extract the downstream remote address.
+
+IP Version Support
+------------------
+The filter supports both IPv4 and IPv6 as addresses. Note that the upstream connection must support
+the version used.
+
+Extra Setup
+-----------
+
+The downstream remote address used will likely be globally routable. By default, packets returning
+from the upstream host to that address will not route through Envoy. The network must be configured
+to forcefully route any traffic whose IP was replicated by Envoy back through the Envoy host.
+
+If Envoy and the upstream are on the same host -- e.g. in an sidecar deployment --, then iptables
+and routing rules can be used to ensure correct behaviour. The filter has an unsigned integer
+configuration,
+:ref:`mark <envoy_api_field_config.filter.listener.original_src.v2alpha1.OriginalSrc.mark>`. Setting
+this to *X* causes Envoy to *mark* all upstream packets originating from this listener with value
+*X*.
+
+We can use the following set of commands to ensure that all ipv4 and ipv6 traffic marked with *X*
```

Comment:
Thank you for putting together these docs @klarose, they're very helpful.
Do you happen to know how the PREROUTING rule below works for local packets?
All the iptables diagrams seem to show no path from local process to PREROUTING so I'm having a hard time understanding how this rule applies to local traffic with the sidecar.

Reply:
It has been a while since I last did this, so I'm a little rusty. As some background, this configuration was inspired by [mmproxy](https://github.com/cloudflare/mmproxy). They don't explain why exactly. :)

I think that it has to do with the fact that there is a separate routing table involved. Perhaps  the system considers transit between different routing tables as transit between different networks?

It could also be that the kernel, seeing that the source IP isn't a local one, treats it as having come from an external network.

I'm sorry I don't have a better answer.

Reply:
That helps, thanks!